net: prefer === to ==

notarseniy opened the pull request:
* Prefer === to == in one condition
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
net
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

TimothyGu commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6555/

vkurchatkin commented:
This is a serious breaking change

jasnell commented:
ah, yes, good point. The `status` change should be ok, but the `options.fd` change is not.

notarseniy commented:
Sorry to ask this, but can someone explain what exactly is the breakdown? It seems that I something misunderstood :(

Trott commented on lib/net.js:
```diff
@@ -146,7 +146,7 @@ function Socket(options) {
     this._handle = createHandle(options.fd);
     this._handle.open(options.fd);
-    if ((options.fd == 1 || options.fd == 2) &&
+    if ((options.fd === 1 || options.fd === 2) &&
```
This could conceivably be a breaking change for some. Since `options.fd` can be provided by the end user, it may be a string. Before this change, passing in `'1'` would have the same effect as `1`. But now it won't. It only affects whether stdout and stderr are blocking or not on Windows. ¯\\_(ツ)_/¯ @nodejs/platform-windows

notarseniy replied:
Ah, got it. Thank you!
Hmm, and what are workarounds? ParseInt'ing? (looks like hack :( ) Leave it as is? (options.fd can be === true, thats not okay?)

Trott replied:
Well, you could just leave `==`. :-D

`options.fd = +options.fd` should make it a number too. There are also bit-shifting things I think that might be more efficient.

But ultimately, I'm not sure this needs to be changed at all.

notarseniy commented:
Reverted `options.fd`-related corrections. Added comment before if-statements for explaining these two non-strict comparisons with link to this PR.

 semver-major label now can be removed. Thanks to @vkurchatkin for pointing this out.

cjihrig commented on lib/net.js:
```diff
@@ -146,6 +146,8 @@ function Socket(options) {
   } else if (options.fd !== undefined) {
     this._handle = createHandle(options.fd);
     this._handle.open(options.fd);
+    // Non-strict equations here for legacy reasons.
```
Instead of saying this and linking to a GitHub issue, could you just explain the legacy reasons.

notarseniy commented:
Expanded commit message; made comment on `options.fd` more better.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6582/

jasnell commented:
Landed in 84c448e
